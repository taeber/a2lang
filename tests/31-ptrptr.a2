asm {
	ORG $800
	JSR main
	JMP $3D0
}

use [
    INIT: sub @ $FB2F,  ; AKA TEXT
    HOME: sub @ $FC58,

    COUT : sub <- [ch: char @ A] @ $FDED,
    CROUT: sub @ $FD8E,
]

var CH: byte @ $24

let Println = sub <- [txt: text @ $06] {
    var [
        i  :int @ Y,
        ch :char,
    ]

    i := 0
    loop if txt_i <> 0 {
        COUT(txt_i)
        i  += 1
    }

    CROUT()
}

let PrintAt = sub <- [txt: text @ $06, col: int @ A] {
    CH := col
    Println(txt)
}

let main = sub {
    INIT()
    HOME()
    PrintAt("Hello, 1979!", col=13)
}

