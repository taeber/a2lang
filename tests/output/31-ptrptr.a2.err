warning: optimized out assigning pointer to itself: Println.txt := PrintAt.txt.
