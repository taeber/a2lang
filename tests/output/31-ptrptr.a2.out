	ORG $800
	JSR main
	JMP $3D0
INIT	EQU $FB2F
HOME	EQU $FC58
COUT	EQU $FDED
CROUT	EQU $FD8E
CH	EQU $24
Println.txt	EQU $06
* COPYBB @Y #$00
Println	LDY #$00
* IFNE (Println.txt),@Y #$00
*   Println._1 Println._2
Println._0	LDA (Println.txt),Y
	CMP #$00
	BNE Println._1
	JMP Println._2
* COPYBB @A (Println.txt),@Y
Println._1	LDA (Println.txt),Y
	JSR COUT
* ADDBB @Y #$01
	INY
	JMP Println._0
Println._2	JSR CROUT
Println._end	RTS
PrintAt.txt	EQU $06
* COPYBB CH @A
PrintAt	STA CH
	JSR Println
PrintAt._end	RTS
main	JSR INIT
	JSR HOME
	LDA #<A2_3
	LDX #>A2_3
	STX PrintAt.txt+1
	STA PrintAt.txt
* COPYBB @A #$0D
	LDA #$0D
	JSR PrintAt
main._end	RTS
Println.ch	HEX 00
A2_3	ASC "Hello, 1979!"
	HEX 00
